replication/OrderedReplication.tla

```diff
@@ -0,0 +1,349 @@
+ *)
+RecvReplicationRequest(r) ==
+    /\ inflightReplicationRequest[r] # {}
+    /\ orderedStoredMessagesPerReplica[r] # {} \* there should be something to replicate.
+    /\ \E msg \in inflightReplicationRequest[r]:
+        /\ inflightReplicationResponse' = [inflightReplicationResponse EXCEPT ![msg.src] = inflightReplicationResponse[msg.src] \union ReplicationResponseMessage(r, orderedStoredMessagesPerReplica, msg) ]
+        /\ inflightReplicationRequest' = [inflightReplicationRequest EXCEPT ![r] = inflightReplicationRequest[r] \ {msg}]
+    /\ UNCHANGED <<replicaStates, blobIdCtr, inflightReplicationMessages, orderedStoredMessagesPerReplica, replicaOffsets, replicaTokens>>
+
+(* Ambry uses a pull model for replication.
+ * In RecvReplicationResponse action, a replica recieves replication response from a peer.
+ * On reciept it checks if each message is present in its local storage log. If not present, then it adds the message, and updates its replica token.
+ *)
+RecvReplicationResponse(r) ==
+    /\ inflightReplicationResponse[r] # {}
+    /\ \E msg \in inflightReplicationResponse[r]:
+        /\ \E message \in msg.messages:
+            /\ IF message.blobmessage = UPDATE_TTL THEN
+                /\ orderedStoredMessagesPerReplica' = [rr \in DOMAIN orderedStoredMessagesPerReplica |-> 
+                    IF ((rr = r) /\ ReplicaCanTakeTtlUpdate(message.blobid, rr, orderedStoredMessagesPerReplica, replicaStates) /\ ~ReplicaHasUpdateTtlMessage(message.blobid, rr, orderedStoredMessagesPerReplica)) THEN
+                        orderedStoredMessagesPerReplica[rr] \union { UpdateTtlRecord(message.blobid, replicaOffsets[rr]) }
+                    ELSE
+                        orderedStoredMessagesPerReplica[rr]
+                    ]
+                ELSE IF message.blobmessage = PUT THEN
+                    /\ orderedStoredMessagesPerReplica' = [ orderedStoredMessagesPerReplica EXCEPT ![r] = 
+                        IF (ReplicaCanTakePut(message.blobid, r, orderedStoredMessagesPerReplica, replicaStates) /\ ~ReplicaHasPutMessage(message.blobid, r, orderedStoredMessagesPerReplica)) THEN
+                            orderedStoredMessagesPerReplica[r] \union { PutRecord(message.blobid, replicaOffsets[r]) }
+                        ELSE 
+                            orderedStoredMessagesPerReplica[r] ]
+                ELSE
+                    /\ orderedStoredMessagesPerReplica' = [rr \in DOMAIN orderedStoredMessagesPerReplica |-> 
+                        IF ((rr = r) /\ ReplicaCanTakeDelete(message.blobid, rr, orderedStoredMessagesPerReplica, replicaStates) /\ ~ReplicaHasDeleteMessage(message.blobid, r, orderedStoredMessagesPerReplica)) THEN
+                            orderedStoredMessagesPerReplica[r] \union { DeleteRecord(message.blobid, replicaOffsets[rr]) }
+                        ELSE
+                            orderedStoredMessagesPerReplica[rr]
+                        ]
+        /\ \E message \in msg.messages:
```

Comment:
So.  This \E message \in msg.messages may refer to a different message then the one picked out in the above clause that looks identical.  Is that ok?  My initial thought looking at this was that we could combine these three for message update/offset update/token update.  Buuuut.... if it's actually desirable they diverge, then maybe this is the way to go?

Reply:
Thanks for pointing this out. I will fix this. I totally missed the fact that `\E` can pick any entry.